From ed4644c6d63935bbdf5b43edaf6d6436c8f76506 Mon Sep 17 00:00:00 2001
From: Shikun Ding <shikun.ding@intel.com>
Date: Tue, 15 Mar 2022 15:43:52 +0800
Subject: [PATCH] Fixed depth not work without set_fmt invoked early

When camera powered on and initialized, using v4l2_ctl command
without formats set option to start streaming will be pending, while
getting the message 'no camera reply from camera processor' in
kernel space.

The reason is state->mux.last_set is pointing to state->motion_t.sensor
, the wrong sensor for streaming depth channel, in ds5_mux_init().
ds5_mux_s_stream() will be invoked after streaming on. The
'sensor->streaming' flag for depth in ds5_configure will not be
set, further skipping the writing 0x1e to override reg. But the override
reg may have the default value, not 0x1e, set by firmware after camera
boots. Then streaming depth will fail.

To fix this, just set state->mux.last_set to correct streaming sersor
on ds5_mux_init stage.

Signed-off-by: Shikun Ding <shikun.ding@intel.com>
---
 drivers/media/i2c/d4xx.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/media/i2c/d4xx.c b/drivers/media/i2c/d4xx.c
index 65a889808..363cd2fa0 100644
--- a/drivers/media/i2c/d4xx.c
+++ b/drivers/media/i2c/d4xx.c
@@ -2777,10 +2777,14 @@ static int ds5_mux_init(struct i2c_client *c, struct ds5 *state)
 	if (ret < 0)
 		goto e_entity;
 
-	if (state->is_rgb)
+	if (state->is_depth)
+		state->mux.last_set = &state->depth.sensor;
+	else if (state->is_rgb)
 		state->mux.last_set = &state->rgb.sensor;
-	else
+	else if (state->is_y8)
 		state->mux.last_set = &state->motion_t.sensor;
+	else
+		state->mux.last_set = &state->imu.sensor;
 
 #ifdef CONFIG_TEGRA_CAMERA_PLATFORM
 	state->mux.sd.dev = &c->dev;
-- 
2.17.1

